### PR TITLE
XGBoost / LightGBM to DAAL converters + examples

### PR DESCRIPTION
**Pull request by RukhovichIV**

### CODE REVIEW

**Comment by ShvetsKS** [comment]
Please move this function implementation to generator/wrapper_gen.py:
you can see `daal_train_test_split` as a reference:
https://github.com/IntelPython/daal4py/blob/master/generator/wrapper_gen.py#L236

File: generator/wrapper_gen.py [hunk 1]
```diff
@@ -245,6 +245,232 @@ def daal_generate_shuffled_indices(idx, random_state):
     c_generate_shuffled_indices(data_or_file(<PyObject*>idx), data_or_file(<PyObject*>random_state))
 
 
+from typing import List, Deque, Dict, Any
+from collections import deque
+from os import remove
+import json
+import numpy as np
```
Comment (ShvetsKS) [inline generator/wrapper_gen.py:252]: move at the beginning of file

File: generator/wrapper_gen.py [hunk 2]
```diff
@@ -245,6 +245,232 @@ def daal_generate_shuffled_indices(idx, random_state):
...
+
+def lgbm_to_daal(model: Any) -> Any:
```
Comment (ShvetsKS) [inline generator/wrapper_gen.py:254]: `lgbm_to_daal` -> `get_gbt_model_from_lgbm` possibly better naming

File: generator/wrapper_gen.py [hunk 3]
```diff
@@ -245,6 +245,232 @@ def daal_generate_shuffled_indices(idx, random_state):
...
+                parentId=parentId, position=position)
+
+            # append children
+            node_stack.append(Node(struct["left_child"], parentId, 0))
+            node_stack.append(Node(struct["right_child"], parentId, 1))
+
+    return mb.model()
+
+
+def xgb_to_daal(booster: Any) -> Any:
```
Comment (ShvetsKS) [inline generator/wrapper_gen.py:343]: `xgb_to_daal` -> `get_gbt_model_from_xgboost` possibly better naming

File: generator/wrapper_gen.py (hunk 1 above)
Comment (RukhovichIV) [reply generator/wrapper_gen.py]: done

File: generator/wrapper_gen.py (hunk 2 above)
Comment (RukhovichIV) [reply generator/wrapper_gen.py]: done

File: generator/wrapper_gen.py (hunk 3 above)
Comment (RukhovichIV) [reply generator/wrapper_gen.py]: done

**Comment by RukhovichIV** [comment]
@Alexander-Makaryev , @PetrovKP , could you review this PR? We need to merge it before code freeze

**Comment by ShvetsKS** [comment]
please update https://github.com/IntelPython/daal4py/blob/master/conda-recipe/meta.yaml#L41 with lgbm and xgboost dependence

**Comment by ShvetsKS** [comment]
also extend test list
https://github.com/IntelPython/daal4py/blob/master/tests/test_examples.py#L199 with new examples, seems (2020,3) should be used

**Comment by ShvetsKS** [comment]
rename gbt_cls_model_create_from_xgboost.py -> gbt_cls_model_create_from_xgboost_batch.py
gbt_cls_model_create_from_lightgbm.py -> gbt_cls_model_create_from_lightgbm_batch.py

**Comment by RukhovichIV** [comment]
@ShvetsKS , thank you, everything is done!

File: generator/wrapper_gen.py [hunk 4]
```diff
@@ -245,6 +251,240 @@ def daal_generate_shuffled_indices(idx, random_state):
...
+    n_features = lgb_model["max_feature_idx"] + 1
+    n_iterations = len(lgb_model["tree_info"]) / lgb_model["num_tree_per_iteration"]
+    n_classes = lgb_model["num_tree_per_iteration"]
+
+    is_regression = False
+    objective_fun = lgb_model["objective"]
+    print("OBJ:   " + objective_fun)
+    if n_classes > 2:
+        if "multiclass" in objective_fun:
+            print("Found multiclass classification")
```
Comment (ShvetsKS) [inline generator/wrapper_gen.py:272]: not sure that pint should be always available, perhaps we should consider verbose mode enabling.

Comment (RukhovichIV) [reply generator/wrapper_gen.py:272]: I forgot to remove it, just like another one in 'example' file.
It'll disappear in the next commit

File: generator/wrapper_gen.py [hunk 5]
```diff
@@ -245,6 +251,240 @@ def daal_generate_shuffled_indices(idx, random_state):
...
+        mb = gbt_clf_model_builder(
+            n_features=n_features, n_iterations=n_iterations, n_classes=n_classes)
+
+    class_label = 0
+    iterations_counter = 0
+    for tree in lgb_model["tree_info"]:
+        if is_regression:
+            tree_id = mb.create_tree(tree["num_leaves"]*2-1)
+        else:
+            tree_id = mb.create_tree(n_nodes=tree["num_leaves"]*2-1, class_label=class_label)
```
Comment (ShvetsKS) [inline generator/wrapper_gen.py:295]: seems `n_nodes=tree["num_leaves"]*2-1` is applicable only for complete binary tree. Do you have some evidence that trees in lgbm model are always compleat?
Seems it would be better to calculate number of nodes by initial look at model.

File: generator/wrapper_gen.py [hunk 6]
```diff
@@ -245,6 +251,240 @@ def daal_generate_shuffled_indices(idx, random_state):
...
+        mb.add_leaf(tree_id=tree_id, response=base_score)
+    else:
+        mb = gbt_clf_model_builder(
+            n_features=n_features, n_iterations=n_iterations, n_classes=n_classes)
+
+    class_label = 0
+    iterations_counter = 0
+    for tree in xgb_model.expandtabs(1).replace(" ", "").split("booster[")[1:]:
+        if is_regression:
+            tree_id = mb.create_tree(len(tree.split("\\n")) - 1)
```
Comment (ShvetsKS) [inline generator/wrapper_gen.py:402]: why we have -1 here?

Comment (RukhovichIV) [reply generator/wrapper_gen.py:402]: beacuse there is one extra-\n in `tree` after split.
It looks like:
\n
node\n
...
node\n

File: generator/wrapper_gen.py (hunk 5 above)
Comment (RukhovichIV) [reply generator/wrapper_gen.py:295]: The fact that the tests are working indirectly proves, that we always have a complete binary tree. When I tried to call `create_tree()` with similar arguments in xgboost convertor, where trees are often incomplete, I always got a memory error while building model in daal. And I never met such errors in LightGBM convertor, so I think, this code might be considered as applicable.

File: generator/wrapper_gen.py (hunk 4 above)
Comment (RukhovichIV) [reply generator/wrapper_gen.py]: all prints are skipped now

File: examples/gbt_cls_model_create_from_lightgbm_batch.py [hunk 7]
```diff
@@ -0,0 +1,92 @@
+#*******************************************************************************
+# Copyright 2014-2020 Intel Corporation
```
Comment (PetrovKP) [inline examples/gbt_cls_model_create_from_lightgbm_batch.py:2]:
```suggestion
# Copyright 2020 Intel Corporation
```

File: examples/gbt_cls_model_create_from_lightgbm_batch.py [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+    lgb_errors_count = np.count_nonzero(np.argmax(lgb_prediction, axis=1) - np.ravel(y_test))
+
+    # Conversion to daal4py
+    daal_model = d4p.get_gbt_model_from_lgbm(lgb_model)
+
+    # daal4py prediction
+    daal_predict_algo = d4p.gbt_classification_prediction(
+        nClasses=5, resultsToEvaluate="computeClassLabels", fptype='float')
+    daal_prediction = daal_predict_algo.compute(X_test, daal_model)
+    daal_errors_count = np.count_nonzero(daal_prediction.prediction - y_test)
```
Comment (PetrovKP) [inline examples/gbt_cls_model_create_from_lightgbm_batch.py:75]: daal4py return y with shape (n, 1), but y_test has shape (n,). Are you sure the operation `-` worked correctly?

File: examples/gbt_cls_model_create_from_lightgbm_batch.py [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+    # LightGBM prediction
+    lgb_prediction = lgb_model.predict(X_test)
+    lgb_errors_count = np.count_nonzero(np.argmax(lgb_prediction, axis=1) - np.ravel(y_test))
+
+    # Conversion to daal4py
+    daal_model = d4p.get_gbt_model_from_lgbm(lgb_model)
+
+    # daal4py prediction
+    daal_predict_algo = d4p.gbt_classification_prediction(
+        nClasses=5, resultsToEvaluate="computeClassLabels", fptype='float')
```
Comment (PetrovKP) [inline examples/gbt_cls_model_create_from_lightgbm_batch.py:73]: Use constant `n_classes` for all functions: for train lgb model and prediction DAAL

File: examples/gbt_cls_model_create_from_lightgbm_batch.py [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+    daal_prediction = daal_predict_algo.compute(X_test, daal_model)
+    daal_errors_count = np.count_nonzero(daal_prediction.prediction - y_test)
+    assert np.absolute(lgb_errors_count - daal_errors_count) == 0
+
+    return (lgb_prediction, lgb_errors_count, np.ravel(daal_prediction.prediction), daal_errors_count, np.ravel(y_test))
+
+
+if __name__ == "__main__":
+    (lgb_prediction, lgb_errors_count, daal_prediction, daal_errors_count, y_test) = main()
+    print("\nLightGBM prediction results (first 10 rows):\n", np.argmax(lgb_prediction[0:10], axis=1))
```
Comment (PetrovKP) [inline examples/gbt_cls_model_create_from_lightgbm_batch.py:83]: Do I understand correctly that there are only 3 lines in a test dataset? Is this significant? Then why are slices in arrays?

File: examples/gbt_cls_model_create_from_lightgbm_batch.py [hunk 11]
```diff
@@ -0,0 +1,92 @@
...
+        'learning_rate':    0.3,
+        'num_class':        5,
+    }
+
+    # Training
+    lgb_model = lgb.train(params, lgb_train, valid_sets=lgb_train)
+
+    # LightGBM prediction
+    lgb_prediction = lgb_model.predict(X_test)
+    lgb_errors_count = np.count_nonzero(np.argmax(lgb_prediction, axis=1) - np.ravel(y_test))
```
Comment (PetrovKP) [inline examples/gbt_cls_model_create_from_lightgbm_batch.py:66]: I see you use translate `np.argmax(lgb_prediction, axis=1)` to labels more one. Maybe use variable?

Comment (PetrovKP) [reply examples/gbt_cls_model_create_from_lightgbm_batch.py:66]: For interest. Did you compare the probabilities?

File: examples/gbt_cls_model_create_from_xgboost_batch.py [hunk 12]
```diff
@@ -0,0 +1,93 @@
+#*******************************************************************************
```
Comment (PetrovKP) [inline examples/gbt_cls_model_create_from_xgboost_batch.py:1]: Same comments from `gbt_cls_model_create_from_lightgbm_batch.py`

File: examples/run_examples.py [hunk 13]
```diff
@@ -80,23 +80,25 @@ def check_device(rule, target):
...
+req_version['adaboost_batch.py'] = (2020,0)
+req_version['brownboost_batch.py'] = (2020,0)
+req_version['dbscan_batch.py'] = (2019,5)
 req_version['decision_forest_classification_batch.py'] = (2019,1)
 req_version['decision_forest_classification_traverse_batch.py'] = (2019,1)
 req_version['decision_forest_regression_batch.py'] = (2019,1)
-req_version['adaboost_batch.py'] = (2020,0)
-req_version['brownboost_batch.py'] = (2020,0)
+req_version['elastic_net_batch.py'] = ((2020,1),(2021,105))
+req_version['gbt_cls_model_create_from_lightgbm_batch.py'] = ((2020, 2), (2021, 109))
```
Comment (PetrovKP) [inline examples/run_examples.py:90]: This feature available from 2020u3 version?
```suggestion
req_version['gbt_cls_model_create_from_lightgbm_batch.py'] = ((2020, 3), (2021, 109))
```

File: generator/wrapper_gen.py [hunk 14]
```diff
@@ -245,6 +251,239 @@ def daal_generate_shuffled_indices(idx, random_state):
     c_generate_shuffled_indices(data_or_file(<PyObject*>idx), data_or_file(<PyObject*>random_state))
```
Comment (PetrovKP) [inline generator/wrapper_gen.py:253]: This file is so huge. Let's move the functionality with converters to another module.

File: generator/wrapper_gen.py [hunk 15]
```diff
@@ -245,6 +251,239 @@ def daal_generate_shuffled_indices(idx, random_state):
...
+
+    is_regression = False
+    objective_fun = lgb_model["objective"]
+    if n_classes > 2:
+        if "multiclass" not in objective_fun:
+            raise TypeError(
+                "multiclass (softmax) objective is only supported for multiclass classification")
+    elif "binary" in objective_fun:  # nClasses == 1
+        # Needed for future verbose
+        # print("Found binary classification")
```
Comment (PetrovKP) [inline generator/wrapper_gen.py:275]: don't forget to delete.

File: generator/wrapper_gen.py [hunk 16]
```diff
@@ -245,6 +251,239 @@ def daal_generate_shuffled_indices(idx, random_state):
...
+        n_classes = 2
+    else:
+        # Needed for future verbose
+        # print("Found regression")
```
Comment (PetrovKP) [inline generator/wrapper_gen.py:279]: don't forget to delete.

File: generator/wrapper_gen.py (hunk 14 above)
Comment (ShvetsKS) [reply generator/wrapper_gen.py:253]: it was discussed with @Alexander-Makaryev, to keep it here like train_test split.
What is your suggestions?

File: examples/gbt_cls_model_create_from_lightgbm_batch.py (hunk 8 above)
Comment (RukhovichIV) [reply examples/gbt_cls_model_create_from_lightgbm_batch.py:75]: y_test also has (n, 1) shape

File: examples/gbt_cls_model_create_from_lightgbm_batch.py (hunk 9 above)
Comment (RukhovichIV) [reply examples/gbt_cls_model_create_from_lightgbm_batch.py]: done

File: examples/gbt_cls_model_create_from_lightgbm_batch.py (hunk 10 above)
Comment (RukhovichIV) [reply examples/gbt_cls_model_create_from_lightgbm_batch.py]: No. There is a 1000 lines in test data set and another 10000 in train.
Yes, it's significant, we don't need to print it all

File: examples/gbt_cls_model_create_from_lightgbm_batch.py (hunk 11 above)
Comment (RukhovichIV) [reply examples/gbt_cls_model_create_from_lightgbm_batch.py]: 1 - done
2 - No, we didn't. I think, that complete accuracy match might be considered as a good indicator

File: examples/run_examples.py (hunk 13 above)
Comment (RukhovichIV) [reply examples/run_examples.py:90]: Yes, but tests are working on version (2020, 2), so our examples are being skipped when we use version (2020, 3)

File: generator/wrapper_gen.py (hunk 15 above)
Comment (RukhovichIV) [reply generator/wrapper_gen.py]: done

File: generator/wrapper_gen.py (hunk 16 above)
Comment (RukhovichIV) [reply generator/wrapper_gen.py]: done too

File: generator/wrapper_gen.py (hunk 6 above)
Comment (RukhovichIV) [reply generator/wrapper_gen.py:393]: I made a mistake. There's no extra-\n at the start of a `tree`, but python's `split()` method also returns the empty lines, so here it works like:
`tree == r"""
node1\n
node2\n
...
noden\n"""`
and `tree.split('\n')` will return `['node1', 'node2', ..., 'noden', '']`, which length is one greater than actual node count.

File: generator/wrapper_gen.py (hunk 14 above)
Comment (PetrovKP) [reply generator/wrapper_gen.py:253]: Create a file `converter_builders.py` and move this content there. From here, for example, import. When all the content is in one file, it is very difficult to read, validate and expand it.

File: examples/run_examples.py (hunk 13 above)
Comment (PetrovKP) [reply examples/run_examples.py:90]: Ok

**Comment by RukhovichIV** [comment]
@PetrovKP , it seems like all your remarks are completed. Can you look at it again and, if everything is ok, merge it?

File: generator/wrapper_gen.py [hunk 17]
```diff
@@ -51,6 +51,13 @@
 import re
 from .wrappers import hpat_types
 
+# from typing import List, Deque, Dict, Any
+# from collections import deque
+# from os import remove
+# import json
+# import re
+
+
```
Comment (ShvetsKS) [inline generator/wrapper_gen.py:60]: delete

Comment (ShvetsKS) [reply generator/wrapper_gen.py:60]: revert changes in this file

**Review by ShvetsKS — APPROVED** [review]

**Review by PetrovKP — APPROVED** [review]
